tests/rptest/tests/topic_recovery_test.py

```diff
@@ -176,17 +176,42 @@ def _validate_partition_last_offset(self):
 
     def _produce_and_verify(self, topic_spec):
         """Try to produce to the topic. The method produces data to the topic and
-        checks that high watermark advanced."""
-        old_hw = None
-        new_hw = None
-        for partition in self._rpk.describe_topic(topic_spec.name):
-            old_hw = partition.high_watermark
-        assert old_hw is not None
+        checks that high watermark advanced. Wait for partition metadata to appear
+        in case of a recent leader election"""
+
+        # Utility to capture the watermark using wait_until as soon as it is not None.
+        class PartitionState:
+            def __init__(self, rpk, topic_name):
+                self.rpk = rpk
+                self.topic_name = topic_name
+                self.hwm = None
+
+            def watermark_is_present(self):
+                for partition_info in self.rpk.describe_topic(self.topic_name):
+                    self.hwm = partition_info.high_watermark
+                return self.hwm is not None
+
+        old_state = PartitionState(self._rpk, topic_spec.name)
+        wait_until(
+            lambda: old_state.watermark_is_present(),
```

Comment:
What about:
```python
lambda : any(self.rpk.describe_topic(self.topic_name))
```
or, if you're paranoid because `partition_info` doesn't have type annotations yet:
```python
lambda: any(map(lambda x: x is not none, self.rpk.describe_topic(self.topic_name))
```

Reply:
Ah.. nevermind. I see you want to keep the hwm for comparison later on.